### PR TITLE
Make Parser data structures const static 

### PR DESCRIPTION
**Pull request by isuruf**

### CODE REVIEW

**Comment by isuruf** [comment]
With this PR,
```
39ms
(x + y - sin(x)/(-4 + z**2) - x**(y**z))**5001
77ms
(x + y - sin(x)/(-4 + z**2) - x**(y**z))**5001
27ms
88ms
```

Master,
```
60ms
(x + y - sin(x)/(-4 + z**2) - x**(y**z))**5001
116ms
(x + y - sin(x)/(-4 + z**2) - x**(y**z))**5001
32ms
89ms
```

**Comment by certik** [comment]
Couple questions:

1. Is the initialization done before main() function gets called in a binary? Does it still take 35ms? If so, I think that's problematic as it will become noticeable. The solution would be to only initialize when the user asks for it.

2. How is it possible that the new parser still got faster? 

Great job!

**Comment by isuruf** [comment]
> 1. Is the initialization done before main() function gets called in a binary? Does it still take 35ms? If so, I think that's problematic as it will become noticeable. The solution would be to only initialize when the user asks for it.

It's undefined in the C++ standard. I think what happens usually is that the initialization happens when the function is called for the first time.

> 2. How is it possible that the new parser still got faster?

There were two lookups per std::map that I've fixed. This was unintentional as the C++ compiler generated an error when `[]` was called on a const std::map, but gives a nice speedup.

File: symengine/parser/parser.cpp [hunk 1]
```diff
@@ -30,50 +30,52 @@ RCP<const Basic> Parser::parse(const std::string &input, bool convert_xor)
...
     if (params.size() == 1) {
-        if (single_arg_functions.find(name) != single_arg_functions.end()) {
-            return single_arg_functions[name](params[0]);
+        auto it1 = single_arg_functions.find(name);
+        if (it1 != single_arg_functions.end()) {
+            return it1->second(params[0]);
         }
-        if (single_arg_boolean_functions.find(name)
-            != single_arg_boolean_functions.end()) {
-            return single_arg_boolean_functions[name](params[0]);
```
Comment (isuruf) [inline symengine/parser/parser.cpp]: @certik, there were two lookups here

**Comment by certik** [comment]
I see. We should figure out exactly when this gets initialized in gcc and clang. If it's the first time the function gets called, that's fine. If it's before main, that would be problematic.

**Comment by isuruf** [comment]
I checked and it's before main.

**Comment by certik** [comment]
Thanks. How long does it take?

We should only do things before main if they are very quick, such as initializing variables to zero and such. If it takes 35ms for every application that links against symengine to start, that prevents to creating command line applications that are fast, among other things. Essentially we become like Python, slow to import. 

I think the solution is to put an if statement in parse, and initialize this at first run. And users who want more control such as our benchmarks can use the Parser class.

**Comment by isuruf** [comment]
With the last commit,
```
26ms
(x + y - sin(x)/(-4 + z**2) - x**(y**z))**5001
54ms
(x + y - sin(x)/(-4 + z**2) - x**(y**z))**5001
18ms
55ms
```
which is even faster. Maybe I'm doing something wrong. Can you run the benchmark for 0f1be78ba, 34620187 and master?

**Comment by isuruf** [comment]
This PR is ready for review. Now the maps are initialized only when the function is called for the first time.

**Comment by certik** [comment]
I'll benchmark it tomorrow. I think the benchmark is timing the initialization process. I would think the code in master makes more sense for the benchmark: it only times the actual parser, not the one time preparation. What do you think? I'll review the rest tomorrow.

**Comment by isuruf** [comment]
`parse('0')` at the beginning takes only `0.019 ms`

**Comment by certik** [comment]
I ran benchmarks. Master:
```
$ rel/benchmarks/parsing 
35ms
(x + y - sin(x)/(-4 + z**2) - x**(y**z))**5001
82ms
(x + y - sin(x)/(-4 + z**2) - x**(y**z))**5001
21ms
100ms
```
and this PR:
```
$ tst/benchmarks/parsing 
parse('0') = 0: 42us
35ms
(x + y - sin(x)/(-4 + z**2) - x**(y**z))**5001
85ms
(x + y - sin(x)/(-4 + z**2) - x**(y**z))**5001
21ms
102ms
```
I ran tihis multiple times. I didn't observe any difference in speed. But the benchmark now looks much cleaner, one just calls `parse()`. So +1 to merge.

**Comment by certik** [comment]
Hold on with the merge yet. Are you sure this doesn't get initialized at startup, but only when `parse()` is called?

**Comment by isuruf** [comment]
Yes, see https://stackoverflow.com/questions/55510/when-do-function-level-static-variables-get-allocated-initialized

**Comment by certik** [comment]
Ok, I tested that example, and even modified it like this:
```c++
void f()
{
        cout << "Function started" << endl;

        static test t("static variable");

        test t2("Local variable");

        cout << "Function executed" << endl;
}
```
and when I run it with g++ 7.4.0, I get:
```
global variable created
local to main created
Program start
Function started
static variable created
Local variable created
Function executed
Local variable destroyed
Program end
local to main destroyed
static variable destroyed
global variable destroyed
```
So it seems static variables get created when they are encountered in the source. Global once get created before main(), but everything else only when needed. Nice!

**Comment by certik** [comment]
Thanks @isuruf, nice addition.

**Pull request merged**